enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/encoding/RaftMessageEncoder.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2002-2018 "Neo4j,"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j Enterprise Edition. The included source
+ * code can be redistributed and/or modified under the terms of the
+ * GNU AFFERO GENERAL PUBLIC LICENSE Version 3
+ * (http://www.fsf.org/licensing/licenses/agpl-3.0.html) with the
+ * Commons Clause, as found in the associated LICENSE.txt file.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * Neo4j object code can be licensed independently from the source
+ * under separate terms from the AGPL. Inquiries can be directed to:
+ * licensing@neo4j.com
+ *
+ * More information is also available at:
+ * https://neo4j.com/licensing/
+ */
+package org.neo4j.causalclustering.messaging.marshalling.v2.encoding;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.MessageToByteEncoder;
+
+import org.neo4j.causalclustering.core.consensus.RaftMessages;
+import org.neo4j.causalclustering.identity.ClusterId;
+import org.neo4j.causalclustering.identity.MemberId;
+import org.neo4j.causalclustering.messaging.NetworkFlushableByteBuf;
+
+public class RaftMessageEncoder extends MessageToByteEncoder<RaftMessages.ClusterIdAwareMessage>
+{
+    @Override
+    protected void encode( ChannelHandlerContext ctx, RaftMessages.ClusterIdAwareMessage decoratedMessage, ByteBuf out ) throws Exception
+    {
+        RaftMessages.RaftMessage message = decoratedMessage.message();
+        ClusterId clusterId = decoratedMessage.clusterId();
+        MemberId.Marshal memberMarshal = new MemberId.Marshal();
+
+        NetworkFlushableByteBuf channel = new NetworkFlushableByteBuf( out );
+        ClusterId.Marshal.INSTANCE.marshal( clusterId, channel );
+        channel.putInt( message.type().ordinal() );
+        memberMarshal.marshal( message.from(), channel );
+
+        message.dispatch( new Handler( memberMarshal, channel ) );
```

Comment:
I think you should be able to instantiate this Handler on creation of the RafeMessageEncoder.